File: src/main/java/ltdjms/discord/shared/DomainError.java
package ltdjms.discord.shared;

import java.util.Objects;

/**
 * Represents domain-level errors that can occur during business operations. This is used as the
 * error type in Result<T, DomainError>.
 */
public record DomainError(Category category, String message, Throwable cause) {

  /** Error categories for classifying different types of domain errors. */
  public enum Category {
    /** Invalid user input or command arguments */
    INVALID_INPUT,
    /** Account balance is insufficient for the operation */
    INSUFFICIENT_BALANCE,
    /** Game token count is insufficient for the operation */
    INSUFFICIENT_TOKENS,
    /** Database or persistence layer failure */
    PERSISTENCE_FAILURE,
    /** Unexpected system failure (bugs, external system issues) */
    UNEXPECTED_FAILURE,
    /** Discord API interaction timeout (3 second limit) */
    DISCORD_INTERACTION_TIMEOUT,
    /** Discord InteractionHook expired (15 minute limit) */
    DISCORD_HOOK_EXPIRED,
    /** Discord message unknown or deleted */
    DISCORD_UNKNOWN_MESSAGE,
    /** Discord API rate limit exceeded */
    DISCORD_RATE_LIMITED,
    /** Discord bot missing required permissions */
    DISCORD_MISSING_PERMISSIONS,
    /** Discord invalid component ID */
    DISCORD_INVALID_COMPONENT_ID,
    /** AI service timeout */
    AI_SERVICE_TIMEOUT,
    /** AI service authentication failed */
    AI_SERVICE_AUTH_FAILED,
    /** AI service rate limited */
    AI_SERVICE_RATE_LIMITED,
    /** AI service unavailable */
    AI_SERVICE_UNAVAILABLE,
    /** AI response empty */
    AI_RESPONSE_EMPTY,
    /** AI response invalid */
    AI_RESPONSE_INVALID,
    /** Prompts directory not found */
    PROMPT_DIR_NOT_FOUND,
    /** Prompt file exceeds size limit */
    PROMPT_FILE_TOO_LARGE,
    /** Prompt file read failed */
    PROMPT_READ_FAILED,
    /** Prompt file has invalid encoding */
    PROMPT_INVALID_ENCODING,
    /** Prompt loading failed */
    PROMPT_LOAD_FAILED
  }

  public DomainError {
    Objects.requireNonNull(category, "category must not be null");
    Objects.requireNonNull(message, "message must not be null");
  }

  /** Creates an invalid input error. */
  public static DomainError invalidInput(String message) {
    return new DomainError(Category.INVALID_INPUT, message, null);
  }

  /** Creates an insufficient balance error. */
  public static DomainError insufficientBalance(String message) {
    return new DomainError(Category.INSUFFICIENT_BALANCE, message, null);
  }

  /** Creates an insufficient tokens error. */
  public static DomainError insufficientTokens(String message) {
    return new DomainError(Category.INSUFFICIENT_TOKENS, message, null);
  }

  /** Creates a persistence failure error. */
  public static DomainError persistenceFailure(String message, Throwable cause) {
    return new DomainError(Category.PERSISTENCE_FAILURE, message, cause);
  }

  /** Creates an unexpected failure error. */
  public static DomainError unexpectedFailure(String message, Throwable cause) {
    return new DomainError(Category.UNEXPECTED_FAILURE, message, cause);
  }
}

File: packages/shared/src/types/domain-error.ts
/**
 * Error categories for classifying different types of domain errors.
 * Matches Java DomainError.Category exactly.
 */
export enum DomainErrorCategory {
  INVALID_INPUT = 'INVALID_INPUT',
  INSUFFICIENT_BALANCE = 'INSUFFICIENT_BALANCE',
  INSUFFICIENT_TOKENS = 'INSUFFICIENT_TOKENS',
  PERSISTENCE_FAILURE = 'PERSISTENCE_FAILURE',
  UNEXPECTED_FAILURE = 'UNEXPECTED_FAILURE',
  DISCORD_INTERACTION_TIMEOUT = 'DISCORD_INTERACTION_TIMEOUT',
  DISCORD_HOOK_EXPIRED = 'DISCORD_HOOK_EXPIRED',
  DISCORD_UNKNOWN_MESSAGE = 'DISCORD_UNKNOWN_MESSAGE',
  DISCORD_RATE_LIMITED = 'DISCORD_RATE_LIMITED',
  DISCORD_MISSING_PERMISSIONS = 'DISCORD_MISSING_PERMISSIONS',
  DISCORD_INVALID_COMPONENT_ID = 'DISCORD_INVALID_COMPONENT_ID',
  AI_SERVICE_TIMEOUT = 'AI_SERVICE_TIMEOUT',
  AI_SERVICE_AUTH_FAILED = 'AI_SERVICE_AUTH_FAILED',
  AI_SERVICE_RATE_LIMITED = 'AI_SERVICE_RATE_LIMITED',
  AI_SERVICE_UNAVAILABLE = 'AI_SERVICE_UNAVAILABLE',
  AI_RESPONSE_EMPTY = 'AI_RESPONSE_EMPTY',
  AI_RESPONSE_INVALID = 'AI_RESPONSE_INVALID',
  PROMPT_DIR_NOT_FOUND = 'PROMPT_DIR_NOT_FOUND',
  PROMPT_FILE_TOO_LARGE = 'PROMPT_FILE_TOO_LARGE',
  PROMPT_READ_FAILED = 'PROMPT_READ_FAILED',
  PROMPT_INVALID_ENCODING = 'PROMPT_INVALID_ENCODING',
  PROMPT_LOAD_FAILED = 'PROMPT_LOAD_FAILED',
  CHANNEL_NOT_ALLOWED = 'CHANNEL_NOT_ALLOWED',
  DUPLICATE_CHANNEL = 'DUPLICATE_CHANNEL',
  INSUFFICIENT_PERMISSIONS = 'INSUFFICIENT_PERMISSIONS',
  CHANNEL_NOT_FOUND = 'CHANNEL_NOT_FOUND',
  DUPLICATE_CATEGORY = 'DUPLICATE_CATEGORY',
  CATEGORY_NOT_FOUND = 'CATEGORY_NOT_FOUND',
  REDEEM_CODE_USED = 'REDEEM_CODE_USED',
  REDEEM_CODE_EXPIRED = 'REDEEM_CODE_EXPIRED',
  REDEEM_CODE_INVALID = 'REDEEM_CODE_INVALID',
}

/**
 * Represents domain-level errors that can occur during business operations.
 * Used as the error type in Result<T, DomainError>.
 * Matches Java DomainError record.
 */
export class DomainError extends Error {
  constructor(
    readonly category: DomainErrorCategory,
    message: string,
    readonly cause?: Error,
  ) {
    // super() MUST be called before any class-level 'this' access in TS strict mode;
    // null checks on constructor parameters are unnecessary because strict TS
    // compilation already prevents null/undefined assignment for typed parameters.
    super(message);
    this.name = 'DomainError';
  }

  // ---- Static factory methods ----

  static invalidInput(message: string): DomainError {
    return new DomainError(DomainErrorCategory.INVALID_INPUT, message);
  }

  static insufficientBalance(message: string): DomainError {
    return new DomainError(DomainErrorCategory.INSUFFICIENT_BALANCE, message);
  }

  static insufficientTokens(message: string): DomainError {
    return new DomainError(DomainErrorCategory.INSUFFICIENT_TOKENS, message);
  }

  static persistenceFailure(message: string, cause?: Error): DomainError {
    return new DomainError(DomainErrorCategory.PERSISTENCE_FAILURE, message, cause);
  }

  static unexpectedFailure(message: string, cause?: Error): DomainError {
    return new DomainError(DomainErrorCategory.UNEXPECTED_FAILURE, message, cause);
  }

  static discordInteractionTimeout(message: string): DomainError {
    return new DomainError(DomainErrorCategory.DISCORD_INTERACTION_TIMEOUT, message);
  }

  static discordHookExpired(message: string): DomainError {
    return new DomainError(DomainErrorCategory.DISCORD_HOOK_EXPIRED, message);
  }

  static discordUnknownMessage(message: string): DomainError {
    return new DomainError(DomainErrorCategory.DISCORD_UNKNOWN_MESSAGE, message);
  }

  static discordRateLimited(message: string): DomainError {
    return new DomainError(DomainErrorCategory.DISCORD_RATE_LIMITED, message);
  }

  static discordMissingPermissions(message: string): DomainError {
    return new DomainError(DomainErrorCategory.DISCORD_MISSING_PERMISSIONS, message);
  }

  static discordInvalidComponentId(message: string): DomainError {
    return new DomainError(DomainErrorCategory.DISCORD_INVALID_COMPONENT_ID, message);
  }

  static aiServiceTimeout(message: string): DomainError {
    return new DomainError(DomainErrorCategory.AI_SERVICE_TIMEOUT, message);
  }

  static aiServiceAuthFailed(message: string): DomainError {
    return new DomainError(DomainErrorCategory.AI_SERVICE_AUTH_FAILED, message);
  }

  static aiServiceRateLimited(message: string): DomainError {
    return new DomainError(DomainErrorCategory.AI_SERVICE_RATE_LIMITED, message);
  }

  static aiServiceUnavailable(message: string): DomainError {
    return new DomainError(DomainErrorCategory.AI_SERVICE_UNAVAILABLE, message);
  }

  static aiResponseEmpty(message: string): DomainError {
    return new DomainError(DomainErrorCategory.AI_RESPONSE_EMPTY, message);
  }

  static aiResponseInvalid(message: string): DomainError {
    return new DomainError(DomainErrorCategory.AI_RESPONSE_INVALID, message);
  }

  static promptDirNotFound(message: string): DomainError {
    return new DomainError(DomainErrorCategory.PROMPT_DIR_NOT_FOUND, message);
  }

  static promptFileTooLarge(message: string): DomainError {
    return new DomainError(DomainErrorCategory.PROMPT_FILE_TOO_LARGE, message);
  }

  static promptReadFailed(message: string, cause?: Error): DomainError {
    return new DomainError(DomainErrorCategory.PROMPT_READ_FAILED, message, cause);
  }

  static promptInvalidEncoding(message: string): DomainError {
    return new DomainError(DomainErrorCategory.PROMPT_INVALID_ENCODING, message);
  }

  static promptLoadFailed(message: string, cause?: Error): DomainError {
    return new DomainError(DomainErrorCategory.PROMPT_LOAD_FAILED, message, cause);
  }

  static channelNotAllowed(message: string): DomainError {
    return new DomainError(DomainErrorCategory.CHANNEL_NOT_ALLOWED, message);
  }

  static duplicateChannel(message: string): DomainError {
    return new DomainError(DomainErrorCategory.DUPLICATE_CHANNEL, message);
  }

  static insufficientPermissions(message: string): DomainError {
    return new DomainError(DomainErrorCategory.INSUFFICIENT_PERMISSIONS, message);
  }

  static channelNotFound(message: string): DomainError {
    return new DomainError(DomainErrorCategory.CHANNEL_NOT_FOUND, message);
  }

  static duplicateCategory(message: string): DomainError {
    return new DomainError(DomainErrorCategory.DUPLICATE_CATEGORY, message);
  }

  static categoryNotFound(message: string): DomainError {
    return new DomainError(DomainErrorCategory.CATEGORY_NOT_FOUND, message);
  }

  static redeemCodeUsed(message: string): DomainError {
    return new DomainError(DomainErrorCategory.REDEEM_CODE_USED, message);
  }

  static redeemCodeExpired(message: string): DomainError {
    return new DomainError(DomainErrorCategory.REDEEM_CODE_EXPIRED, message);
  }

  static redeemCodeInvalid(message: string): DomainError {
    return new DomainError(DomainErrorCategory.REDEEM_CODE_INVALID, message);
  }
}
